File: C++/maximum-subarray-sum-after-multiplier.cpp
// Time:  O(n)
// Space: O(1)

// dp, kadane's algorithm
class Solution {
public:
    long long maxSubarraySum(vector<int>& nums, int k) {
        static const auto& NEG_INF = numeric_limits<int64_t>::min();
        enum {INIT, MULT, DIV, DONE};

        int64_t result = NEG_INF;
        vector<int64_t> dp(4, NEG_INF);
        for (const int64_t x : nums) {
            dp = {
                max(dp[INIT], static_cast<int64_t>(0)) + x,
                max({dp[MULT], dp[INIT], static_cast<int64_t>(0)}) + x * k,
                max({dp[DIV], dp[INIT], static_cast<int64_t>(0)}) + x / k,
                max({dp[DONE], dp[MULT], dp[DIV], static_cast<int64_t>(0)}) + x,
            };
            result = max(result, ranges::max(dp));
        }
        return result;
    }
};

File: Python/maximum-subarray-sum-after-multiplier.py
# Time:  O(n)
# Space: O(1)

# dp, kadane's algorithm
class Solution(object):
    def maxSubarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        def sign(x):
            return 1 if x >= 0 else -1

        NEG_ING = float("-inf")
        INIT, MULT, DIV, DONE = range(4)
        result = NEG_ING
        dp = [NEG_ING]*4
        for x in nums:
            dp[:] = [
                max(dp[INIT], 0)+x,
                max(dp[MULT], dp[INIT], 0)+x*k,
                max(dp[DIV], dp[INIT], 0)+sign(x)*(abs(x)//k),
                max(dp[DONE], dp[MULT], dp[DIV], 0)+x,
            ]
            result = max(result, max(dp))
        return result
